File: instrumentserver/keysightAWG/digitizer test/DaqBufferExample.cpp
// Example using DAQbufferPoolConfig
//
#include <iostream>
#include <Windows.h>
#include <conio.h>
#include "SD_AIN.h"
#include "sd_constants.h"
#include "SD_Module.h"
#include "SD_AOU.h"
#include "SD_WAVE.h"
using namespace std;

void myCallback(SD_Object* SDobject, int eventNumber, void* buffer0, int buffer0Points,
   void* buffer1, int buffer1Points, void* userObject, int status);

int main()
{
   int chassis = 1;
   int digSlot = 5;
   int digChannel = 1;

   int status, i;
   int channels[] = { 1 };
   int numberOfChannels = 1;

   double fullScale = 2;
   int pointsPerCycle = 20;
   int cycles = 1000;
   int delay = 0;

   int AIN_IMPEDANCE_HZ = 0;
   int AIN_IMPEDANCE_50 = 1;
   int AIN_COUPLING_DC = 0;
   int AIN_COUPLING_AC = 1;

   int timeOut = 10000;

   int BUFFER_SIZE = pointsPerCycle * cycles / 2;

   short *buffer1 = new short[BUFFER_SIZE];
   short *buffer2 = new short[BUFFER_SIZE];


   cout << "Enter chassis number: " << endl;
   cin >> chassis;
   char a;
   if (cin.fail()) {
      cout << "Invalid chassis" << endl;     
      getch();
      return 0;
   }

   cout << "Enter digitizer slot number: " << endl;
   cin >> digSlot;
   if (cin.fail()) {
      cout << "failed" << endl;
      getch();
      return 0;
   }
   

   SD_AIN *dig = new SD_AIN();

   if ((status = dig->open("M3102A", chassis, digSlot)) < 0)
   {
      cout << "Could not open digitizer: " << status << endl;
      getch();
      return 0;
   }
   
   dig->DAQflush(digChannel);
   
   dig->channelInputConfig(digChannel, fullScale, AIN_IMPEDANCE_50, AIN_COUPLING_DC);
   dig->DAQconfig(digChannel, pointsPerCycle, cycles, delay, SWHVITRIG);
      

   //Callback setup - not available in Python
   callbackEventPtr cb = callbackEventPtr(&myCallback);
   dig->DAQbufferPoolConfig(digChannel, buffer1, BUFFER_SIZE, timeOut, cb, NULL);
   dig->DAQbufferAdd(digChannel, buffer2, BUFFER_SIZE);
   

   dig->DAQstart(digChannel);
   cout << "Setup done." << endl;
   
   //Trigger or wait for all cycles
   for (i = 0; i < cycles; i++) 
   {
      Sleep(10);
      dig->DAQtrigger(digChannel);
   }
   
   
   cin >> status;
   cout << "done" << endl;
   dig->close();
   delete(buffer1);
   delete(buffer2);

   return 0;

}


//This function gets called every time the provided buffer gets full
//Here, one can save data to file, plot etc.
void myCallback(SD_Object* SDobject, int eventNumber, void* buffer0, int buffer0Points,
   void* buffer1, int buffer1Points, void* userObject, int status) 
{
   //As an example, we simply print out the acquired data to the console
   int i; 
   for (i = 0; i < buffer0Points; i++) 
   {
      cout << ((short*)buffer0)[i] << " " << std::flush; 
   }
   
   cout << endl;  
}

File: instrumentserver/keysightAWG/digitizer test/DaqBufferExample.py
import os
import sys
import time
import matplotlib.pyplot as plt
import numpy as np
sys.path.append('C:\Program Files (x86)\Keysight\SD1\Libraries\Python')
import keysightSD1 as key



pointsPerCycle = 10000
captureCycles = 1


# This variable defines how many total transactions will happen
# (pointsPerCycle * captureCycles / transfers) must be an even integer 
transfers = 1


chassis = 0
digSlot = 3

digName = "M3102A"


digChannels = [1, 2, 3, 4]
digScale = 2
captureDelay = 0


dig = key.SD_AIN()
error = dig.openWithSlotCompatibility(digName, chassis, digSlot, key.SD_Compatibility.KEYSIGHT)
if error < 0:
   print("Digitizer open error: {}".format(error))
   dig.close()
   exit()

error = dig.triggerIOconfig(key.SD_TriggerDirections.AOU_TRG_IN)
if error < 0:
      print('Error triggerIOconfig {}'.format(key.SD_Error.getErrorMessage(error)))

for i in range(len(digChannels)):   

   error = dig.DAQtriggerExternalConfig(digChannels[i], key.SD_TriggerExternalSources.TRIGGER_EXTERN, key.SD_TriggerBehaviors.TRIGGER_RISE, key.SD_SyncModes.SYNC_NONE)
   if error < 0:
      print('Error DAQtriggerExternalConfig - {}'.format( key.SD_Error.getErrorMessage(error)))

   error = dig.DAQflush(digChannels[i])
   if error < 0:
      print('Error DAQflush {}'.format(error))

   error = dig.channelInputConfig(digChannels[i], digScale, key.AIN_Impedance.AIN_IMPEDANCE_50, key.AIN_Coupling.AIN_COUPLING_DC)
   if error < 0:
      print('Error channelInputConfig {}'.format(error))

   error = dig.DAQconfig(digChannels[i], pointsPerCycle, captureCycles, captureDelay, key.SD_TriggerModes.EXTTRIG)
   if error < 0:
      print('Error DAQconfig {}'.format(error))

   error = dig.DAQbufferPoolConfig(digChannels[i], int(pointsPerCycle * captureCycles / transfers))
   if error < 0:
      print('Error DAQbufferPoolConfig {}'.format(error))

   error = dig.DAQstart(digChannels[i])
   if error < 0:
         print('Error DAQstart {}'.format(error))   


# Add code to either trigger digitizer or wait for first few cycles
#time.sleep(1)


data = []

for i in range(len(digChannels)):
   data.append(dig.DAQbufferGet(digChannels[i]))

for j in range(1, transfers):
   for i in range(len(digChannels)):
      temp = dig.DAQbufferGet(digChannels[i])      
      data[i] = np.concatenate((data[i], temp), axis=0)     #only here for data plotting purposes, this concatenation can be done afterwards for efficiency


for i in range(len(digChannels)):
   plt.plot(data[i], label=str(i+1))
plt.legend()
plt.show()


for i in range(len(digChannels)):
   error = dig.DAQbufferPoolRelease(digChannels[i])
   if error < 0:
      print('Error DAQbufferPoolRelease {}'.format(error))


dig.close()

print("Done.")